packages/compass-aggregations/src/stores/store.ts

```diff
@@ -258,6 +259,28 @@ export function activateAggregationsPlugin(
     maxTimeMSChanged(preferences.getPreferences().maxTimeMS || null)
   );
 
+  addCleanup(
+    workspaces.registerTabCloseHandler?.(async function (nextTab) {
```

Comment:
What is the advantage of conflating this into one handler with 4 states? wouldn't it be easier to reason with 2 handlers just returning true or false?

```
workspaces.registerTabCloseHandler(() => {
    if (!store.getState().isModified) return true;
    return await showConfirmation({
        title: 'Your aggregation contains unsaved changes',
        description:
          'You will loose all unsaved changes to the pipeline you are currently building',
        buttonText: 'Close tab',
        variant: 'danger',
      });
}),
workspaces.registerTabReplaceHandler(() => {
    return !store.getState().isModified;
});
```

Reply:
Separate statuses allow us to handle multiple handlers registered in the same tab.

Imagine both Aggergations and Documents sub tabs of a Collection workspace tab registered the handlers that do different checks. If handlers just return true / false and we stop checking after the first handler returned true, you might accidentally loose content that the other handler is checking. If you just start calling all of them, you might run the risk of showing multiple popups. Hence a way to communicate explicitly that user did an action and the checks don't need to go further.

We can drop extra statuses if we're okay with either
* Just not having custom per-workspace logic. We won't be able to conditionally show confirmation modal on close in that case unless we move the state of Aggergations plugin to workspaces somehow, which I'd try to avoid
* Changing the interface here completely to syncronously return something that workspaces plugin can use to make the decision whether or not an aggregated confirmation modal should be shown. This would require workspaces to be more aware of how to handle different results returned from this method (like we wouldn't want to show one if users were just clicking around the tab)
 
The second approach would be okay if we really don't want to have different statuses returned here the way they are right now, but we would either way need to provide something more specific than true / false to allow workspaces to make the call on what to do, I just tried to separate the states required for making a call on what to do with the tab from the logic that derives this informatino from user input that might be plugin-specific.

Reply:
I'm a big fan of "Just not having custom per-workspace logic".

>  If handlers just return true / false and we stop checking after the first handler returned true

That makes sense, but we can still split replace and close right?

Also not to be super nit-picky but I think this is just an implementation detail, we can also do something like:

```
canReplace: () => boolean
canClose: () => true | () => boolean
```

like that we can call all the registered `canClose` handlers (which can be sync) to understand if a tab can be closed right away, and then call the first or last function as confirmation.

Not as a proof of anything, but Intuitively the jsdoc of "canClose" would just be: "determine if a tab can be closed. the handler must return `true` if the tab can be closed, or a confirmation function in case it can't.", and you don't necessarily need to know anything about the underlying mechanics. Right now if you write the handler wrong you can screw things up.

Reply:
Totally, as I mentioned, if design is okay with just having generic "You have some changes somewhere, are you sure you want to close the tab?" instead of being more specific, we can totally change that, remove the flexible interface, register separate handlers for separate cases, and move the actual confirmation logic to the workspace plugin, it was just not what was initially assumed as what we want in the current designs we have as a reference

Reply:
As far as i understand we don't want custom logic for replace, the 2 options are:

- just use the exact same that is used for "close"
- use do something generic like clicks

I like both of them, what i don't particularly find worth is "let's define 100 conditions to replace tabs", it would both be annoying to maintain and unpredictable for users.

Also, as I said, unless canClose === canReplace all of the times, I find collapsing replace and close logic in one handler quite complex to reason with. So also if we just say `canClose === canReplace` we would probably remove that issue.

Reply:
I don't think we can say `canClose === canReplace` in all cases if our generic way of determining whether or not the tab can't be closed is just clicking around it, there will be too many false positives. And we'd still need to be able to register different checks for canClose in different places of the app in that case to prompt users only if actually meaningful content was edited and can get lost

Reply:
Updated to use canClose / canReplace sync response in latest commit https://github.com/mongodb-js/compass/pull/5952/commits/3c4cc7e732fc97f4dd013878e491ba5919bf242a

Reply:
@mcasimir updated once more to two separate handlers as interface, do you mind taking another look to confirm that this matches your expectations after our last discussion?

Reply:
(I'll resolve this one, but feel free to start new threads if the new implementation doesn't match the expectation)